**SA402 &#x25aa; Dynamic and Stochastic Models &#x25aa; Fall 2023**

# How to Win at Monopoly

## Background

<img src="monopoly.jpg" width="60%">

* Here's a very, very basic guide to the rules of Monopoly.

* There are 40 **board positions**, numbered 0 to 39, some of which correspond to **properties**. [See the tables at the end of this notebook.](#tables)


* All players start at position 0 ("Go"). At each turn, a player rolls 2 six-sided dice and moves according to the sum of their values.


* If the player lands on a property that is not owned by another player, he or she has the opportunity to **purchase** it. If the property is already owned by another player, then he or she has to **pay rent** to the owner.


* If the player lands on "Chance" or "Community Chest," then the player must randomly draw a "Chance" or "Community Chest" card, which tells the player to collect money, pay money, or go to a different board position. [See the tables at the end of this notebook for the distributions of the "Chance" and "Community Chest" cards.](#tables)

## Objective

* If we know which properties are landed on the most, this could help us devise a good strategy for which properties to buy. 


* We will determine which properties are landed on the most by modeling movement on the board as a Markov chain, in which **each board position corresponds to a state**.


* _For the Monopoly veterans:_ 
    - We're going to ignore any "rolling doubles" rules (especially the one where 3 double rolls sends a player to Jail). 
    - In addition, we're going to assume that a player leaves Jail after 1 turn (that is, going to Jail just moves the player's position). 
    - It turns out that these assumptions affect the accuracy of our results only slightly.

## Let's analyze this Markov chain

* In the same folder as this notebook, there is a file `monopoly.py` that defines the transition probability matrix for the Monopoly Markov chain. 


* Open the file and take a look. The variable `P` is the transition probability matrix $\mathbf{P}$.


* We can import the variable `P` from `monopoly.py` as follows:  

In [1]:
from monopoly import P

* We can inspect `P` as follows:

In [2]:
P

matrix([[0.01215278, 0.        , 0.02430556, 0.05555556, 0.09375   ,
         0.12152778, 0.13888889, 0.07291667, 0.13888889, 0.11111111,
         0.08333333, 0.06597222, 0.03819444, 0.        , 0.        ,
         0.01041667, 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.01041667,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.01215278, 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.01041667],
        [0.00868056, 0.        , 0.        , 0.02777778, 0.06423611,
         0.09201389, 0.11111111, 0.06076389, 0.16666667, 0.13888889,
         0.11111111, 0.09201389, 0.06423611, 0.02777778, 0.        ,
         0.00868056, 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.00868056,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.00868056, 0.        , 

* Note that `P` is a [NumPy](https://numpy.org/) [matrix](https://numpy.org/doc/stable/reference/generated/numpy.matrix.html), which represents a matrix, roughly speaking, as a list of rows. 


* For example, the matrix
$$
A = \begin{bmatrix}
    1 & 2 & 3\\
    4 & 5 & 6\\
    7 & 8 & 9
    \end{bmatrix}
$$
is represented as

```
matrix([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])
```

### Step 1

* Let's first try to understand what some of these transition probabilities are.


* Consider $p_{14,18}$, the transition probability between Virginia Avenue (14) and Tennessee Avenue (18). We can access the relevant entry of the NumPy matrix `P` using

    ```python
    P[14, 18]
    ```
    
    <br>Give it a try below.

In [3]:
# Solution
P[14, 18]

0.08333333333333333

What is the value of $p_{14,18}$? Briefly explain how this transition probability is derived.

_Write your notes here. Double-click to edit._

_Solution._ $p_{14,18}$ is equal to the probability that we roll a 4, which is $\frac{3}{36} \approx 0.0833$.

### Step 2

* Now consider the transition probability between New York Avenue (19) and Jail (30). According to the transition probability matrix, what is its value? Briefly explain why.


* _Hint_. [See the tables at the end of this notebook.](#tables) Remember that you can go to Jail from the Chance #2 board position (22).

In [4]:
# Solution
P[19, 30]

0.059027777777777776

_Write your notes here. Double-click to edit._

_Solution._ $p_{19,30}$ is equal to the probability that we roll an 11, *or* we roll a 3 and then go to Jail by drawing a Chance card. Therefore, $p_{19,30} = \frac{2}{36} + \frac{2}{36} \big( \frac{1}{16} \big) \approx 0.0590$.

### Step 3

* Compute $\mathbf{P}^{1000}$. Recall that in Python, we write $x^y$ as `x**y`.

In [5]:
# Solution
P**1000

matrix([[0.0310836 , 0.02160563, 0.01910468, 0.02203332, 0.02366539,
         0.0285958 , 0.02297443, 0.01025009, 0.02348758, 0.02330611,
         0.02299686, 0.02736561, 0.02628363, 0.02393293, 0.02480968,
         0.02784624, 0.02786918, 0.02576628, 0.02918729, 0.03065187,
         0.02863573, 0.02816088, 0.01214307, 0.02723292, 0.03177897,
         0.02885491, 0.02709774, 0.02683874, 0.02816546, 0.02597727,
         0.03585242, 0.02675852, 0.02615659, 0.02379903, 0.02495268,
         0.02431131, 0.01012184, 0.02194854, 0.02192537, 0.02647182],
        [0.0310836 , 0.02160563, 0.01910468, 0.02203332, 0.02366539,
         0.0285958 , 0.02297443, 0.01025009, 0.02348758, 0.02330611,
         0.02299686, 0.02736561, 0.02628363, 0.02393293, 0.02480968,
         0.02784624, 0.02786918, 0.02576628, 0.02918729, 0.03065187,
         0.02863573, 0.02816088, 0.01214307, 0.02723292, 0.03177897,
         0.02885491, 0.02709774, 0.02683874, 0.02816546, 0.02597727,
         0.03585242, 0.02675852, 

* Does $\mathbf{P}^{1000}$ have a particular structure? Why do you think it looks the way it does?

_Write your notes here. Double-click to edit._

_Solution._ All the rows are identical, because after 1000 steps, the Markov chain has reached steady state.

* Using your computed $\mathbf{P}^{1000}$, give an educated guess on which states are transient and which states are recurrrent.

_Write your notes here. Double-click to edit._

_Solution_. $p_{ii}^{(1000)} > 0$ for all states $i = 0, 1, \dots, 39$, so an educated guess would be that the limiting probabilities $p_{ii}^{(\infty)} > 0$ for all states $i = 0, 1, \dots, 39$ as well. As a result, _all states_ are recurrent. 

### Step 4

* Recall that we can compute the steady-state probabilities by solving the following system of equations:
    $$
    \begin{aligned}
    \mathbf{\pi}^{\top} \mathbf{P} & = \mathbf{\pi}^{\top}\\
    \mathbf{\pi}^{\top} \mathbf{1} & = 1
    \end{aligned}
    $$

    We can rewrite this system of equations as
    $$
    \begin{aligned}
    (\mathbf{I} - \mathbf{P})^{\top} \mathbf{\pi} & = \mathbf{0}\\
    \mathbf{1}^{\top} \mathbf{\pi} & = 1
    \end{aligned}
    $$

    or equivalently,

    $$
    \underbrace{\left[ \begin{array}{c}
        (\mathbf{I} - \mathbf{P})^{\top} \\
        1 \cdots 1
    \end{array} \right]}_{\mathbf{A}}
    \mathbf{\pi}
    = 
    \underbrace{\left[ \begin{array}{c}
        0\\
        \vdots\\
        0\\
        1
    \end{array} \right]}_{\mathbf{b}}
    $$


* To set up and solve this system, we will use some NumPy functions.


* First, let's import NumPy.

In [6]:
import numpy as np

* To form the $\mathbf{A}$ matrix above, we need to stack $(\mathbf{I} - \mathbf{P})^{\top}$ and $[1 \cdots 1 ]$ vertically.


* In NumPy:
    - The $n \times n$ identity matrix is `np.eye(n)`.
    - The transpose of matrix `M` is `M.T`.
    - The $m \times n$ matrix of 0s is `np.zeros((m, n))`.
    - The $m \times n$ matrix of 1s is `np.ones((m, n))`.
    - We can vertically stack matrices `R` and `S` using `np.vstack([R, S])`.


* Using this information, define a NumPy matrix `A` that represents the matrix $\mathbf{A}$ defined above.

In [7]:
# Solution
A = np.vstack([(np.eye(40) - P).T, np.ones((1,40))])

* Next, define a NumPy matrix `b` that represent the vector $\mathbf{b}$ defined above.

In [8]:
# Solution
b = np.vstack([np.zeros((40,1)), np.ones((1,1))])

* Finally, we can solve the system $\mathbf{A} \pi = \mathbf{b}$ using the method `np.linalg.lstsq` like this:

In [9]:
pi, *other_info = np.linalg.lstsq(A, b, rcond=None)

# Display pi, transposed so that it fits on our screens
pi.T

array([[0.0310836 , 0.02160563, 0.01910468, 0.02203332, 0.02366539,
        0.0285958 , 0.02297443, 0.01025009, 0.02348758, 0.02330611,
        0.02299686, 0.02736561, 0.02628363, 0.02393293, 0.02480968,
        0.02784624, 0.02786918, 0.02576628, 0.02918729, 0.03065187,
        0.02863573, 0.02816088, 0.01214307, 0.02723292, 0.03177897,
        0.02885491, 0.02709774, 0.02683874, 0.02816546, 0.02597727,
        0.03585242, 0.02675852, 0.02615659, 0.02379903, 0.02495268,
        0.02431131, 0.01012184, 0.02194854, 0.02192537, 0.02647182]])

* What is the long-run fraction of time that the game is at Reading Railroad? How about Boardwalk?

_Write your notes here. Double-click to edit._

_Solution._ The long-run fraction of time that the game is at Reading Railroad is 0.0286. The long-run fraction of time that the game is at Boardwalk is 0.0265. 

* Compare the $\pi$ you found to $\mathbf{P}^{1000}$. What do you see? Is this what you expected? Why?

_Write your notes here. Double-click to edit._

_Solution._ $\pi$ is equal to the rows of $\mathbf{P}^{1000}$. This indicates that after 1000 steps, the Markov chain has reached steady state.

### Step 5

* In the long run, what are the 5 most visited board positions?

* _Hint._ In NumPy, the following code will take a column vector `x`, transpose it, and return the **indices** that sort the elements of `x` from lowest to highest.

    ```python
    np.argsort(x.T)
    ```

In [10]:
# Solution
np.argsort(pi.T)

array([[36,  7, 22,  2,  1, 38, 37,  3,  6, 10,  9,  8,  4, 33, 13, 35,
        14, 34, 17, 29, 32, 12, 39, 31, 27, 26, 23, 11, 15, 16, 21, 28,
         5, 20, 25, 18, 19,  0, 24, 30]])

_Write your notes here. Double-click to edit._

_Solution._ The five most visited board positions are 30 (Jail), 24 (Illinois Avenue), 0 (Go), 19 (New York Avenue), and 18 (Tennessee Avenue).

### Step 6

* For those of you familiar with Monopoly, you know that the most visited properties might not be the most valuable, because the rent you can charge differs from property to property.


* Run the cell below, which imports a NumPy column vector called `rent` from `monopoly.py`. 


* This vector contains the rents you can charge on each property if you own 1 house on that property.
    * _For the Monopoly veterans:_ we will ignore utilities and railroads and assume the rents on those properties is \$0, since the renting rules are a bit more complicated for those properties.

In [11]:
from monopoly import rent

* We can `rent` as follows:

In [12]:
# Show rent vector, transposed
rent.T

matrix([[ 0.,  2.,  0.,  4.,  0.,  0.,  6.,  0.,  6.,  8.,  0., 10.,  0.,
         10., 12.,  0., 14.,  0., 14., 16.,  0., 18.,  0., 18., 20.,  0.,
         22., 22.,  0., 24.,  0., 26., 26.,  0., 28.,  0.,  0., 35.,  0.,
         50.]])

* So, for example, if you owned one house on Boardwalk (39), you would receive \$50 every time a player lands on Boardwalk.

* How would you use the vector `rent` with the steady state probability vector `pi` you found above to find the __expected one-house rent per dice roll__ for each property?

* _Hint._ In NumPy, you can multiply two vectors `x1` and `x2` component-wise with:

    ```python
    np.multiply(x1, x2)
    ```

In [13]:
# Solution
expected_rent = np.multiply(rent, pi)

* Which 5 properties have the highest expected one-house rent per dice roll? Based on this, what can you say about which properties are better or worse to own?

In [14]:
# Solution
np.argsort(expected_rent.T)

matrix([[ 0, 33, 25, 35, 22, 36, 20, 38, 17, 15, 30, 12, 28, 10,  7,  5,
          4,  2,  1,  3,  6,  8,  9, 13, 11, 14, 16, 18, 23, 19, 21, 27,
         26, 29, 24, 32, 31, 34, 37, 39]])

_Write your notes here. Double-click to edit._

_Solution._ The five properties with the highest expected one-house rents are 39 (Boardwalk), 37 (Park Place), 34 (Pennsylvania Avenue) , 31 (Pacific Avenue), 32 (North Carolina Avenue).

### Step 7

* Suppose we want to incorporate the "rolling doubles" rules. In particular:
    - If the player rolls a double, they roll again after taking their turn. If they roll three doubles in a row, then they go to jail.
    - A player stays in Jail for three turns, or until he or she rolls a double.


* How would you change the Markov chain to incorporate these rules?

_Write your notes here. Double-click to edit._

_Solution._ Expand the state space: use states $(x, y)$, where $x$ represents the board positions ($x = 0, \dots, 39$), and $y$ represents the number of consecutive doubles rolled so far ($y = 0, 1, 2)$.

## Tables
<a id="tables"></a>

### Board positions

| State | Board Position                      | State | Board Position        | State | Board Position                            |
| ----- | ----------------------------------- | ----- | --------------------- | ----- | ----------------------------------------- |
| 0     | Go                                  | 14    | Virginia Avenue       | 27    | Ventnor Avenue                            |
| 1     | Mediterranean Avenue                | 15    | Pennsylvania Railroad | 28    | Water Works                               |
| 2     | Community Chest #1                  | 16    | St. James Place       | 29    | Marvin Gardens                            |
| 3     | Baltic Avenue                       | 17    | Community Chest #2    | 30    | Go to Jail (we use this as being in Jail) |
| 4     | Income Tax                          | 18    | Tennessee Avenue      | 31    | Pacific Avenue                            |
| 5     | Reading Railroad                    | 19    | New York Avenue       | 32    | North Carolina Avenue                     |
| 6     | Oriental Avenue                     | 20    | Free Parking          | 33    | Community Chest #3                        |
| 7     | Chance #1                           | 21    | Kentucky Avenue       | 34    | Pennsylvania Avenue                       |
| 8     | Vermont Avenue                      | 22    | Chance #2             | 35    | Short Line                                |
| 9     | Connecticut Avenue                  | 23    | Indiana Avenue        | 36    | Chance #3                                 |
| 10    | Jail (we use this as visiting Jail) | 24    | Illinois Avenue       | 37    | Park Place                                |
| 11    | St. Charles Place                   | 25    | B& O Railroad         | 38    | Luxury Tax                                |
| 12    | Electric Company                    | 26    | Atlantic Avenue       | 39    | Boardwalk                                 |
| 13    | States Avenue                       |       


### Chance card distribution

| Destination                          | Probability |
| ------------------------------------ | ----------- |
| Go (0)                               | 1/16        |
| Reading Railroad (5)                 | 1/16        |
| St.~Charles Place (11)               | 1/16        |
| Illinois Avenue (24)                 | 1/16        |
| Jail (30)                            | 1/16        |
| Boardwalk (39)                       | 1/16        |
| Nearest utility (forward direction)  | 1/16        |
| Nearest railroad (forward direction) | 1/16        |
| 3 spaces back                        | 1/16        |
| Stay put                             | 7/16        |

### Community Chest card distribution

| Destination | Probability |
| ----------- | ----------- |
| Go (1)      | 1/16        |
| Jail (31)   | 1/16        |
| Stay put    | 14/16       |